worlds/Files.py

```diff
@@ -190,21 +204,193 @@ def get_source_data_with_cache(cls) -> bytes:
+        ("apply_bsdiff4", ["delta.bsdiff4"])
+    ]
+
+    def __init__(self, *args: Any, patched_path: str = "", **kwargs: Any) -> None:
+        super(APDeltaPatch, self).__init__(*args, **kwargs)
+        self.patched_path = patched_path
+
     def write_contents(self, opened_zipfile: zipfile.ZipFile):
+        self.write_file("delta.bsdiff4",
+                        bsdiff4.diff(self.get_source_data_with_cache(), open(self.patched_path, "rb").read()))
         super(APDeltaPatch, self).write_contents(opened_zipfile)
-        # write Delta
-        opened_zipfile.writestr("delta.bsdiff4",
-                                bsdiff4.diff(self.get_source_data_with_cache(), open(self.patched_path, "rb").read()),
-                                compress_type=zipfile.ZIP_STORED)  # bsdiff4 is a format with integrated compression
-
-    def read_contents(self, opened_zipfile: zipfile.ZipFile):
-        super(APDeltaPatch, self).read_contents(opened_zipfile)
-        self.delta = opened_zipfile.read("delta.bsdiff4")
-
-    def patch(self, target: str):
-        """Base + Delta -> Patched"""
-        if not self.delta:
-            self.read()
-        result = bsdiff4.patch(self.get_source_data_with_cache(), self.delta)
-        with open(target, "wb") as f:
-            f.write(result)
+
+
+class APTokenTypes(IntEnum):
+    WRITE = 0
+    COPY = 1
+    RLE = 2
+    AND_8 = 3
+    OR_8 = 4
+    XOR_8 = 5
+
+
+class APTokenMixin:
+    """
+    A class that defines functions for generating a token binary, for use in patches.
+    """
+    tokens: List[
+        Tuple[int, int,
+        Union[
+            bytes,  # WRITE
+            Tuple[int, int],  # COPY, RLE
+            int  # AND_8, OR_8, XOR_8
+        ]]] = []
+
+    def get_token_binary(self) -> bytes:
+        """
+        Returns the token binary created from stored tokens.
+        :return: A bytes object representing the token data.
+        """
+        data = bytearray()
+        data.extend(len(self.tokens).to_bytes(4, "little"))
+        for token_type, offset, args in self.tokens:
+            data.append(token_type)
+            data.extend(offset.to_bytes(4, "little"))
+            if token_type in [APTokenTypes.AND_8, APTokenTypes.OR_8, APTokenTypes.XOR_8]:
+                data.extend(int.to_bytes(1, 4, "little"))
+                data.append(args)
+            elif token_type in [APTokenTypes.COPY, APTokenTypes.RLE]:
+                data.extend(int.to_bytes(4, 4, "little"))
+                data.extend(args[0].to_bytes(4, "little"))
+                data.extend(args[1].to_bytes(4, "little"))
+            else:
+                data.extend(len(args).to_bytes(4, "little"))
+                data.extend(args)
+        return data
+
+    def write_token(self, token_type: APTokenTypes, offset: int, data: Union[bytes, Tuple[int, int], int]):
+        """
+        Stores a token to be used by patching.
+        """
+        self.tokens.append((token_type, offset, data))
```

Comment:
here we could `if __debug__: assert isinstance(...)` based on token_type

Reply:
why `if __debug__:` ?
Isn't it the same without that?

```python3
a = compile("assert 1", "", "exec", optimize=0)
b = compile("if __debug__: assert 1", "", "exec", optimize=0)

print(a == b)  # True

a_o = compile("assert 1", "", "exec", optimize=1)

print(a == a_o)  # False

b_o = compile("if __debug__: assert 1", "", "exec", optimize=1)

print(a_o == b_o)  # True

c = compile("", "", "exec", optimize=0)

print(a_o == c)  # True
```

Reply:
Not if you have the following:
```python
if token_type in [...]:
    assert ...
elif token_type in [...]:
    assert ...
...
```
Pretty sure the `if`s can not be dropped because the `in` operator could have side-effects.

Reply:
There is a nicer approach to it though:
```python
token_arg_type: Dict[APTokenTypes, type] = {
    APTokenTypes.COPY: tuple
    ...
}
assert isinstance(data, token_arg_type[token_type])
```
which does not need `if __debug__`